### PR TITLE
Example implemented using NetworkAPI

### PR DESCRIPTION
**Pull request by dkeeney**
This is a new example program.

This seems very slow.  Can someone check my parameters?

### CODE REVIEW

File: src/examples/hotgym_NetworkAPI/hotgym_napi.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+  try {
+
+    std::cout << "initializing\n";
+
+    Network net;
+
+    // Declare the regions to use
+    std::string encoder_parameters = "{size: " + std::to_string(DIM_INPUT) + ", activeBits: 4, radius: 0.5, seed: 2019 }";
+    std::shared_ptr<Region> region1 =  net.addRegion("region1", "RDSERegion", encoder_parameters);
```
Comment (breznak) [inline src/examples/hotgym_NetworkAPI/hotgym_napi.cpp]: nit, use more descriptive/intuitive region instances' names, ie "inputRegionRDSE, spRegion, spGlobalRegion, tmRegion" ?

File: src/examples/hotgym_NetworkAPI/hotgym_napi.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    std::shared_ptr<Region> region2a = net.addRegion("region2a", "SPRegion", "{columnCount: " + std::to_string(COLS) + "}");
```
Comment (breznak) [inline src/examples/hotgym_NetworkAPI/hotgym_napi.cpp]: ad slow: you have SP with local inh connected, so I think this slows things down. I think it's fine to use just one (global) in the example.

File: src/examples/hotgym_NetworkAPI/hotgym_napi.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    std::shared_ptr<Region> region2b = net.addRegion("region2b", "SPRegion", "{columnCount: " + std::to_string(COLS) + ", globalInhibition: true}");
```
Comment (breznak) [inline src/examples/hotgym_NetworkAPI/hotgym_napi.cpp]: regarding parameters, these setings use SPs' etc default params, I think it'd be nice to use: 
- I think the best would be if you also add SensorRegion to show reading a CSV file, so to replicate `hotgym.py` (uses the real hotgym data, also then use its params)

File: src/examples/hotgym_NetworkAPI/README.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
+﻿# C++ example using Network API
+The program `hotgym_napi` is an example of an app using the Network API tools available in the htm.core library.  In this example we generate a sin wave with some noise as the input.  This is passed to an encoder to turn that into SDR format.  This is passed to two instances of SpatialPooler (SP), one is configured for local inhibition and one for global inhibition.  The output of the SP for global inhibition is passed on to the temporalMemory (TM) algorithm.  The output of the TM can be written to a file so that it can be plotted.
```
Comment (breznak) [inline src/examples/hotgym_NetworkAPI/README.md]: nit, sine wave :)

File: src/examples/hotgym_NetworkAPI/hotgym_napi.cpp (hunk 1 above)
Comment (dkeeney) [reply src/examples/hotgym_NetworkAPI/hotgym_napi.cpp]: Ok, I was thinking about that too.

File: src/examples/hotgym_NetworkAPI/hotgym_napi.cpp (hunk 2 above)
Comment (dkeeney) [reply src/examples/hotgym_NetworkAPI/hotgym_napi.cpp]: Ok, I was just trying to duplicate what the other hotgym example was doing.
With the debugger can tell that it is the SP that was really slow...both to initialize and to run.

File: src/examples/hotgym_NetworkAPI/hotgym_napi.cpp (hunk 3 above)
Comment (dkeeney) [reply src/examples/hotgym_NetworkAPI/hotgym_napi.cpp]: Ok, I did not even think to pattern it after hotgym.py.  I assumed it was the same as HelloSPTP.cpp.  I will check that out.

File: src/examples/hotgym_NetworkAPI/README.md (hunk 4 above)
Comment (dkeeney) [reply src/examples/hotgym_NetworkAPI/README.md]: :-)

**Comment by dkeeney** [comment]
It looks like I need to make some more Regions.
- Predictor
- AnomalyLikelihood
- Date Encoder
- maybe Plot    (using matplotlib perhaps)

And I like the idea of setting up the parameters in a table.  What if I used a Yaml table?
For the plot we can call out to Python but I cannot be sure that it will be available. But if it is, it would be nice to have a region that runs it.

There is also a problem in knowing where the .csv file exists.  The program cannot assume it is in the same folder as the executable...most likely not.  Have to think about that.

**Comment by breznak** [comment]
> It looks like I need to make some more Regions.
    Predictor

probably yes, also SDRClassifier. 

>    AnomalyLikelihood

this should be already available in `TM.anomaly` (does the TM region expose that field?)

>    Date Encoder
    maybe Plot (using matplotlib perhaps)

would be nice to have. Date enc might see some larger changes, but who knows when.. (there's an issue for that) 

> And I like the idea of setting up the parameters in a table. What if I used a Yaml table?

do you mean a YAML file with the params? My ideal goal would be that Algorithms, and Regions support additional constructors with a (yaml/json) config file, so you can change settings w/o recompiling. 

> For the plot we can call out to Python but I cannot be sure that it will be available. But if it is, it would be nice to have a region that runs it. 

What about a py Region for plotting if python is avail, and we can sort out plotting with no py later? 

> There is also a problem in knowing where the .csv file exists. The program cannot assume it is in the same folder as the executable...most likely not. Have to think about that.

this should be a minor issue. Expect the file path given as arg. And default the data can be copied by cmake to the bin/ folder (?)

**Comment by dkeeney** [comment]
>  available in TM.anomaly (does the TM region expose that field?)

Yes, anomaly is there but how do I get AnomalyLikelihood from it?  I will probably need to add some code someplace to get that.

> do you mean a YAML file with the params?

yes.  I could pass a Yaml File into the addRegion call, or by expanding this a little, pass one yaml file to Network and it could define the regions to add as well as the parameters to apply to them.

> Expect the file path given as arg. And default the data can be copied by cmake to the bin/ folder (?)

That was my thinking.  Although when the program runs it knows only the default folder and that can be any place.  It does not know where the repository is or even if one exists.  In linux, you use cmake to do an install step before running the programs so more than likely, if you built from sources, the executable will be in the build/Release/bin folder.   At least on Windows, if you do a debug build within MSVC it does not do the install step first so it executes it from someplace down in build/scripts/?? so in this case it will not know where to find the csv file.  anyway, I will think of something.

**Comment by dkeeney** [comment]
So, it looks like I need to set this PR aside until I can do the following:
1) Predictor; port algorithm and create Region
2) SDRClassifier; create Region
3) Date Encoder; port algorithm and create Region
4) AnomalyLikelihood; (modify TMRegion to provide)
5) Parameters in yaml that can be passed to Network to configure all regions
6) A new .py Region for plots callable from C++ and C# apps

**Comment by dkeeney** [comment]
Oh, the SDRClassifier is actually the Predictor.  Very confusing.

**Comment by breznak** [comment]
> Yes, anomaly is there but how do I get AnomalyLikelihood from it?

TM has a param `anomalyMode` (or sth) that sets the type of anomaly you get from the field TM.anomaly. 

> So, it looks like I need to set this PR aside until I can do the following

how about wrapping this up to make it HelloSPTP alternative, as originaly intended, merge, and then extend to the hotgym.py functionality? 

steps 1,4 are imho incorrect, 5,6 optional (would suit a separate PR each). 

> Oh, the SDRClassifier is actually the Predictor. Very confusing.

it's in the same file, if you split it into standalone, it'd be fine. 
Predictor uses Classifiers and builds atop them.

**Comment by dkeeney** [comment]
It appears that the HelloSPTP.cpp example has nothing to do with hotgym so the hotgym.cpp is a bad name.  Perhaps it should be called sine_wave or something.

So you think I should also create a HelloSPTP version using NetworkAPI in addition to the real hotgym.  I guess I could.

**Comment by breznak** [comment]
> It appears that the HelloSPTP.cpp example has nothing to do with hotgym 

well, it does. It uses the same processing pipeline: SP-TM-Anomaly/prediction. ATM it's just not yet as complete as the py version. 
The c++ is also used to benchmark our performance in a "typical workload". 

> So you think I should also create a HelloSPTP version using NetworkAPI in addition to the real hotgym

no, I was suggesting to merge this as-is (=similar to HelloSPTP) for now, and then build other NetworkAPI additions on it, to get closer to the hotgym.py functionality

**Comment by dkeeney** [comment]
> no, I was suggesting to merge this as-is (=similar to HelloSPTP) for now, and then build other NetworkAPI additions on it, to get closer to the hotgym.py functionality

They are so different, I think it better to have two separate examples.
I will merge the one I have with some adjustments.

**Comment by dkeeney** [comment]
The anomaly always comes out as 1.
It runs so slow that I had to cut back on the iterations and length of SDR to make it a reasonable demo.  But this probably does not iterate enough times to learn anything.

I still have the SP (local) in the program.  Since that is broken, let me try taking that out and see how much faster it runs.

**Comment by dkeeney** [comment]
yes, it runs much better without SP (local).  SP (global) takes about 10 seconds to initialize (Windows, debug mode) but I guess that is ok.
I am going to leave it with SP(local) commented out rather than removing it because I assume that it will eventually be fixed.

**Comment by breznak** [comment]
> I still have the SP (local) in the program. Since that is broken,

SP with local inhibition is not broken, it's just slow. The usefulness us now questionable, but in a way the produced SDRs are of a better quality. We should try speeding it up.

File: src/examples/napi_sine/README.md [hunk 5]
```diff
@@ -1,5 +1,12 @@
 ﻿# C++ example using Network API
-The program `hotgym_napi` is an example of an app using the Network API tools available in the htm.core library.  In this example we generate a sin wave with some noise as the input.  This is passed to an encoder to turn that into SDR format.  This is passed to two instances of SpatialPooler (SP), one is configured for local inhibition and one for global inhibition.  The output of the SP for global inhibition is passed on to the temporalMemory (TM) algorithm.  The output of the TM can be written to a file so that it can be plotted.
+The program `sine_napi` is an example of an app using the Network API tools 
+available in the htm.core library.  In this example we generate a sine wave 
+with some noise as the input.  This is passed to an encoder to turn that 
+into SDR format.  This is passed to two instances of SpatialPooler (SP),
```
Comment (breznak) [inline src/examples/napi_sine/README.md]: nit, SP makes the SDR format, encoder does "just" binary representation of whatever raw inputs. Also, do you want to mention the 2 SP instances, when only one is actually used?

File: src/examples/napi_sine/napi_sine.cpp [hunk 6]
```diff
@@ -36,9 +37,14 @@ int main(int argc, char* argv[]) {
...
-  //  Runtime arguments:  hotgym_napi [epochs [filename]]
+  std::string encoder_params   = "{size: " + std::to_string(DIM_INPUT) + ", activeBits: 4, radius: 0.5, seed: 2019 }";
+//  std::string sp_local_params  = "{columnCount: " + std::to_string(COLS) + "}";
+  std::string sp_global_params = "{columnCount: " + std::to_string(COLS) + ", globalInhibition: true}";
+  std::string tm_params        = "{activationThreshold: 9, cellsPerColumn: " + std::to_string(CELLS) + "}";
+
+  //  Runtime arguments:  napi_sine [epochs [filename]]
   if(argc >= 2) {
-    EPOCHS = std::stoi(argv[1]);         // number of iterations (default 5000)
+    EPOCHS = std::stoi(argv[1]);         // number of iterations (default 500)
```
Comment (breznak) [inline src/examples/napi_sine/napi_sine.cpp]: 500 iters seems really too little, don't think TM will learn anything. Why is this (is it?) slower? The HelloSPTM runs 5000 just fine.

File: src/examples/napi_sine/README.md (hunk 5 above)
Comment (dkeeney) [reply src/examples/napi_sine/README.md]: > do you want to mention the 2 SP instances

Oh, yes. I should do that.  Actually, I think I will take out the second SP entirely.  This is kind of a Hello World example and should be as simple as possible.

File: src/examples/napi_sine/napi_sine.cpp (hunk 6 above)
Comment (dkeeney) [reply src/examples/napi_sine/napi_sine.cpp]: > 500 iters seems really too little

yeah, EPOCHS default is actually 5000, its the comment that is wrong.   Fixed it.

Comment (breznak) [reply src/examples/napi_sine/napi_sine.cpp]: and still, after 5000 steps the predictions are bad? (the results should be comparable, actually even identical if you make the same sine wave and fixed random as with the Algo version. )

**Comment by dkeeney** [comment]
Thanks for approving this, but I think this should remain as a Hello World type example and I should build a second example that matches the hotgym.py example.  Both would have value.

I was not trying to match the results of HelloSPTP in this PR but perhaps I should.  I am computing the sine differently and I may have different parameters.  So let me see if I can make them more the same.

The fact that both are bad is also a problem.  This is probably a user's first exposure to htm.core, at least that is its purpose and it should have at least reasonable results.  We control the example and we should pick an example that shows what htm can do.  To not do so is embarrassing.  That is bugger than what should be done in this PR so lets put that in another PR but it should be done.

**Comment by breznak** [comment]
> I was not trying to match the results of HelloSPTP in this PR but perhaps I should. 

if not 1:1 deterministic results, similar quality of results would be good! 

> The fact that both are bad is also a problem. This is probably a user's first exposure to htm.core, at least that is its purpose and it should have at least reasonable results

ok, as a separate issue, we should look at the results & params, it's true that avg 0.4 anomaly on a sine wave in 5000 steps isn't stellar. (I don't know if the random noice isn't too strong there etc) 

This PR, feel free to merge as is :+1:

**Comment by dkeeney** [comment]
> f not 1:1 deterministic results, similar quality of results would be good!

At the moment I am not even close.
The input from the encoder is now the same but SP and TM are very different and anomaly and likelyhood are way off.   These should all match exactly since I am calling the same functions with the same parameters.  So I have something now quite right.  I will keep working on it.

**Comment by dkeeney** [comment]
Found some things that needed correcting:
- Found that some of the defaults in the SP and TM algorithms were not the same as the defaults in the Regions. Remember that if you change the defaults in an algorithm be sure to also change it in its corresponding Region.

- found that computation of sine starts with 0.01 in HelloSPTP and not 0.0. Changed napi_sine to match.

- found that in HelloSPTP, `anLikelihood.anomalyProbability(an)` was not being assigned to anlikely.  That is why it was always 0.

- also found that in HelloSPTP, the _ARCH_DETERMINISTIC checks were never made.  Did NOT look into why.

**Comment by dkeeney** [comment]
Additional things we might want to consider:
- adding a reset a the beginning of each sine wave cycle so the algorithm knows when its the beginning of the pattern.
- iterate the sine wave using degrees rather than radians so the cycle begins on an even sample boundary.

**Comment by dkeeney** [comment]
@breznak  I will need another review when you find time.

I will be traveling starting tomorrow, returning probably Sat.  Happy Thanksgiving everyone.

**Comment by dkeeney** [comment]
```
benchmark_hotgym
    Epoch = 5000
    Anomaly = 0.77451
    Anomaly (avg) = 0.411894
    Anomaly (Likelihood) = 0.88831

napi_sine
     Result after 5000 iterations.
      Anomaly             = 0.77451
      Anomaly(avg)        = 0.411894
      Anomaly(Likelihood) = 0.88831
```

File: src/examples/hotgym/HelloSPTP.cpp [hunk 7]
```diff
@@ -154,7 +154,7 @@ EPOCHS = 2; // make test faster in Debug
       avgAnomOld_ = avgAnom10.getCurrentAvg(); //update
     }
     tAnLikelihood.start();
-    anLikelihood.anomalyProbability(an); //FIXME AnLikelihood is 0.0, probably not working correctly
+    anLikely = anLikelihood.anomalyProbability(an); //FIXME AnLikelihood is 0.0, probably not working correctly
```
Comment (breznak) [inline src/examples/hotgym/HelloSPTP.cpp]: thanks for fixing this mistake! 
Nit, the FIXME note could now go away

File: src/examples/hotgym/HelloSPTP.hpp [hunk 8]
```diff
@@ -16,7 +16,7 @@ class BenchmarkHotgym {
 public:	
   Real64 run(
     UInt EPOCHS = 5000,
-    bool useSPlocal=true, //can toggle which (long running) components are tested, default all
+    bool useSPlocal=false, //can toggle which (long running) components are tested, default all
```
Comment (breznak) [inline src/examples/hotgym/HelloSPTP.hpp]: nit, please keep the local inh SP enabled for the test, it's one of the places ensuring its code runs compatible.

File: src/htm/regions/RDSERegion.cpp [hunk 9]
```diff
@@ -47,18 +48,23 @@ namespace htm {
           resolution:  {type: Real32, default: "0.0"},
           category:    {type: Bool,   default: "false"},
           seed:        {type: UInt32, default: "0"},
-          sensedValue: {type: Real64, default: "0.0", access: ReadWrite }},
+          noise:       {description: "amount of noise to add to the output SDR. 0.01 is 1%",
```
Comment (breznak) [inline src/htm/regions/RDSERegion.cpp]: nice idea to have the noice included in the encoder :+1: 
Maybe for later, it'd be useful to have a "NoisyRegion", so that it can be applied to/after any layer. That is similar to dropout PR #535

File: src/htm/regions/TMRegion.cpp [hunk 10]
```diff
@@ -336,7 +343,7 @@ Spec *TMRegion::createSpec() {
           NTA_BasicType_UInt32,          // type
           1,                             // elementCount
           "",                            // constraints
-          "8",                           // defaultValue
+          "10",                          // defaultValue
```
Comment (breznak) [inline src/htm/regions/TMRegion.cpp]: so the defaults now match the Algo version (?) It's good that thanks to this example, we've validated that people using NetworkAPI get the same quality out of the box :+1:

File: src/test/unit/regions/TMRegionTest.cpp [hunk 11]
```diff
@@ -213,7 +213,7 @@ TEST(TMRegionTest, testLinking) {
   std::shared_ptr<Region> region1 = net.addRegion("region1", "VectorFileSensor",parameters);
   std::shared_ptr<Region> region2 = net.addRegion("region2", "SPRegion", "{dim: [2,10]}");
   std::shared_ptr<Region> region3 = net.addRegion("region3", "TMRegion",
-                                        "{activationThreshold: 9, cellsPerColumn: 5}");
+                                        "{activationThreshold: 11, cellsPerColumn: 5}");
```
Comment (breznak) [inline src/test/unit/regions/TMRegionTest.cpp]: no problem, but why these changes?

File: src/htm/regions/TMRegion.cpp (hunk 10 above)
Comment (dkeeney) [reply src/htm/regions/TMRegion.cpp]: This is a test to confirm that this field can be changed.  Could have used any field.  I had to use a value of 11 rather than 8 because it had to be larger than the minimum which is now 10.

File: src/htm/regions/RDSERegion.cpp (hunk 9 above)
Comment (dkeeney) [reply src/htm/regions/RDSERegion.cpp]: Good idea.

File: src/examples/hotgym/HelloSPTP.hpp (hunk 8 above)
Comment (dkeeney) [reply src/examples/hotgym/HelloSPTP.hpp]: > please keep the local inh SP enabled

Ok, I had turned it off when comparing with napi_sine.  Forgot to turn it back on.  Good catch.

File: src/htm/regions/TMRegion.cpp (hunk 10 above)
Comment (dkeeney) [reply src/htm/regions/TMRegion.cpp]: Yes, I changed all NetworkAPI defaults to match the Algorithm defaults.  This required a manual eye-ball compare.  We don't have a way to write a test that can check if the defaults are the same in all cases.

File: src/test/unit/regions/TMRegionTest.cpp (hunk 11 above)
Comment (dkeeney) [reply src/test/unit/regions/TMRegionTest.cpp]: >  why these changes?

This test is to make sure that a field can be set to something other than the default.  I had to change the values for activationThreshold in this tesst because it had to be larger than the minimum which is now 10.
